### PR TITLE
AO EOF fix and protection

### PR DESCRIPTION
**Pull request by ashwinstar**
PR contains fix for issue and also enhanced protection to prevent data corruption in any case.

SnapshotNow MUST be used during insert flow to fetch the latest EOF value from
aoseg table. Earlier usage of ActiveSnapshot in AO insert flow caused data
corruption as potentially could read incorrect or stale EOF value from
aoseg. The specific scenario this happens when entry from AppendOnlyHash table
gets evicted out. It invalidates (zeros out) the latestWriteXid and hence
check for usedByConcurrentTransaction cannot be performed. So, any transaction
later inserting data to same AO table, if it has aquired the snapshot and has
latestWriteXid listed in its in-progress distributed transaction list, gets
the same segfile to write. Based on DTM visibility rules using ActiveSnapshot
current transaction will not read the EOF value from previous inserted
transaction as its listed in its in-progress list and hence will overwrite the
data but SnapshotNow will see the latest and make sure to append.

### CODE REVIEW

**Comment by hlinnaka** [comment]
Hmm. If the entry is evicted from the hash table, so that usedByConcurrentTransaction() fails to notice that the segment was in fact just inserted to by another transaction, don't you potentially get incorrect query results in concurrent transactions, too?

BTW, I'm getting rid of the whole hash table in PR #790, so this might become obsolete, or need a different fix, after that.

**Comment by ashwinstar** [comment]
On Thu, Aug 11, 2016 at 2:42 AM, Heikki Linnakangas <
notifications@github.com> wrote:

> Hmm. If the entry is evicted from the hash table, so that
> usedByConcurrentTransaction() fails to notice that the segment was in fact
> just inserted to by another transaction, don't you potentially get
> incorrect query results in concurrent transactions, too?
>
Nope usedByConcurrentTransaction is only used for assigning which segment
file to insert data too not during reads / selects. So, doesn't affect
query result of concurrent transaction. Selects continue to use regular
snapshot only insert uses SnapshotNow to read aoseg entry. The entry gets
evicted out only post transaction writing to it has marked it not in use
and distributed committed which means definitely has completed work on
segment.
Only in one extremely remote case the query fails if all this is true

   - concurrent transaction is Serializable and
   - entry gets evicted out and
   - performs insert to table
   - followed by select <--- this select will fail to provide data

Was planning to fix that remote case by performing the reverse-map from
local xid fetched from aoseg entry to distributed xid, if latestWriteXid is
0 in cache for serializable insert transaction due to cache eviction and
the use the same during usedByConcurrentTransaction. But then since anyways
its so remote case only with serializable transaction and with your PR we
may potential change the design, decided to just put in backlog for now.

BTW, I'm getting rid of the whole hash table in PR #790
> <https://github.com/greenplum-db/gpdb/pull/790>, so this might become
> obsolete, or need a different fix, after that.
>
Yes, will look into the same again along with the PR #790 but anyways the
fix is needed for both 4.3_STABLE and master in current state. Plus,
irrecpective the change seems to be useful
1] because makes code cleaner
2] reduces diff between AO and CO code
3] also reduces amount of work done during visibility checking of aoseg
entries as not needed to evaluate in-progress array once you have exclusive
access

**Comment by ashwinstar** [comment]
@hlinnaka any further comments, planning to push this in.

**Comment by hlinnaka** [comment]
No objections. It's not a complete fix, but it's a step in the right direction.

On 12 August 2016 20:11:12 EEST, Ashwin Agrawal <notifications@github.com> wrote:
>@hlinnaka any further comments, planning to push this in.
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/greenplum-db/gpdb/pull/1034#issuecomment-239504167

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

File: src/include/cdb/cdbappendonlyam.h [hunk 1]
```diff
@@ -343,7 +343,7 @@ extern void appendonly_fetch_detail(
 	AppendOnlyFetchDesc aoFetchDesc, 
 	AppendOnlyFetchDetail *aoFetchDetail);
 extern void appendonly_fetch_finish(AppendOnlyFetchDesc aoFetchDesc);
-extern AppendOnlyInsertDesc appendonly_insert_init(Relation rel, Snapshot appendOnlyMetaDataSnapshot, int segno, bool update_mode);
+extern AppendOnlyInsertDesc appendonly_insert_init(Relation rel, int segno, bool update_mode);
```
Comment (asimrp) [inline src/include/cdb/cdbappendonlyam.h]: Good that we are not passing snapshots around anymore.

File: src/test/isolation/specs/ao-insert-eof.spec [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+step "s1commit"	{ COMMIT; }
+
+session "s2"
+step "s2begin"   { BEGIN ISOLATION LEVEL SERIALIZABLE; }
+step "s2select"  { SELECT count(*) from appendonly_eof; }
+step "s2insert"	 { INSERT INTO appendonly_eof SELECT * FROM generate_series(1, 10); }
+step "s2commit"  { COMMIT; }
+
+
+permutation "s1begin" "s1setguc" "s1insert" "s2begin" "s2select" "s1commit" "s2insert" "s2commit" "s2select"
```
Comment (asimrp) [inline src/test/isolation/specs/ao-insert-eof.spec]: Love the isolation framework!

File: src/backend/utils/misc/guc_gp.c [hunk 3]
```diff
@@ -1680,6 +1681,15 @@ struct config_bool ConfigureNamesBool_gp[] =
 		false, NULL, NULL
 	},
 
+	{
+		{"test_AppendOnlyHash_eviction_vs_just_marking_not_inuse", PGC_SUSET, DEVELOPER_OPTIONS,
```
Comment (asimrp) [inline src/backend/utils/misc/guc_gp.c]: Nitpick - can we name the GUC as just "test_AppendOnlyHash_evict_unused_entry"?

**Comment by asimrp** [comment]
LGTM :+1:

**Pull request merged**